### ISSUE TITLE
[$500] Chat - Map request position is displayed wrong in workspace

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#29607]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.84.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1697195763235749

## Action Performed:
1. Open the app
2. Open any workspace chat which already has 1 map request pending
3. Send some messages
4. Click on green plus, request money->distance
5. Select any start and finish distance and click Next
6. Select workspace report open in step 2 and complete the process
7. It will open the workspace chat, wait until map is loaded and observe that map request position switches back to previous position instead of displaying as latest message

## Expected Result:
Map request should be displayed as latest message in workspace report if no messages are sent after raising the request

## Actual Result:
Map request is not displayed as latest message in workspace report even if no messages are sent after raising the request (position issue is rectified on revisit)

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/1eeadc8e-68ba-4c7c-953c-50c3321be0a6

https://github.com/Expensify/App/assets/93399543/47a3dfdd-c76c-483d-af11-9e74469ecbc8

https://github.com/Expensify/App/assets/93399543/c3ae9ee9-14ff-44b6-b981-fc9770e65a8a

</details>

<details>
<summary>MacOS: Desktop</summary>

 

https://github.com/Expensify/App/assets/93399543/c5e225cc-7229-43ff-a994-96c08e897ce8

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01753ab0226b6e4e39</li>
        <li>Upwork Job ID: 1713204618313187328</li>
        <li>Last Price Increase: 2023-10-14</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @roryabraham</details>

**Comment by jliexpensify** [comment]
Hi @roryabraham - could you confirm this is an `Internal` issue based off [this](https://github.com/Expensify/App/issues/29607#issuecomment-1764918591)? Thanks in advance!

**Comment by Victor-Nyagudi** [comment]
I also uncovered signs of a back end issue. 

If you inspect the `reportActions` prop passed to `ReportScreen`, you'll notice that the report preview is initially at the right place (first in the array) but jumps to the position after the most recently sent comment(s) .

Furthermore, the `created` property of the report preview's object is set to a time before the latest comment, possibly referencing the previous distance request instead of the previous message. 

Take a look.

<details>

<summary>Video of report preview jumping to different spot</summary>

https://github.com/Expensify/App/assets/79470910/4ab9cb3c-2a90-4a21-89e7-be617c342f42

</details> 

It's worth noting that reloading the page or switching to a different tab and back renders the map position in the right place, and this is most likely (as far as I can tell) due to `OpenReport` API call happening and we receive the correct report preview timestamp. 

I suspect `createDistanceRequest` API call is culprit here because if you repeat the reproduction steps offline, the distance request is rendered at the correct position, even if you do go back online. Additionally, from what I could gather when looking at Onyx, the correct `created` timestamp is stored in local storage for the new distance request. 

https://github.com/Expensify/App/blob/75a2cc9aac7a49cf3ea55041909b5b85f7c0f6cb/src/libs/actions/IOU.js#L617

Take a look at what happens offline. 

<details>

<summary>Video of trying to reproduce the bug while offline</summary>

https://github.com/Expensify/App/assets/79470910/2a6b046a-8905-43c1-a5dc-8fd297e94747

</details> 

Speaking of the `created` timestamp, it's worth noting that the report preview's timestamp is updated to the most recent distance request, however, the latest distance request is manually merged in `getMoneyRequestInformation` instead of using Onyx. 

Take a look at the comment before this happens. 

https://github.com/Expensify/App/blob/75a2cc9aac7a49cf3ea55041909b5b85f7c0f6cb/src/libs/actions/IOU.js#L509-L520

I tried setting the `created` property of `optimisticTransaction` back to the optimistic version instead of the existing transaction out of curiosity, and the latest distance request did render after the most recent message as intended but as a standalone money request instead of as part of the report preview. 

```javascript
optimisticTransaction = {
    ...optimisticTransaction,
    ..existingTransaction,
    created: optimisticTransaction.created // Added 
};
```

In conclusion, I think it would be good to not only check on what's going on in the back end but to also see if what @tgolen was working on in this [issue](https://github.com/Expensify/App/issues/25417) could make a difference.

**Comment by roryabraham** [comment]
Working on reproducing this but currently wrestling with my dev environment after updating to Sonoma

**Comment by Christinadobrzyn** [comment]
hi there - could you confirm if this is related or will be fixed with this? https://github.com/Expensify/App/issues/30147

**Comment by roryabraham** [comment]
I don't think this is quite the same, no. My understanding of these two issues:

- This one – the `createDistanceRequest` API request is returning the wrong timestamp for the IOU action
- The other one - the IOU preview is showing in the right place, but the previous IOU action is not being shown at all. I think we should be showing a plain text message in the report history with the original IOU when it was requested and just the preview should move

**Comment by roryabraham** [comment]
Got my dev environment prettymuch all working again, but still have no update here.

**Comment by roryabraham** [comment]
Working on this in https://github.com/Expensify/Auth/pull/9038

**Comment by jliexpensify** [comment]
Being worked on!

**Comment by jliexpensify** [comment]
Not overdue! @roryabraham should we make this a `Weekly` instead?

**Comment by jliexpensify** [comment]
Not overdue

**Comment by jliexpensify** [comment]
Not overdue

**Comment by situchan** [comment]
Same

**Comment by roryabraham** [comment]
I must apologize for the lack of communication on this issue. The PR I was working on to fix this was deployed to production about a month ago, but the issue persists. Going to bring this to slack to discuss

**Comment by roryabraham** [comment]
Actually, found a simple solution to fix this, so opened a PR: https://github.com/Expensify/Web-Expensify/pull/40008

**Comment by jliexpensify** [comment]
Going an Upworks job to sort our Reporting payment for @dhanashree-sawant

**Comment by dhanashree-sawant** [comment]
Thanks @jliexpensify, I have accepted the offer.

**Comment by jliexpensify** [comment]
Paid and job closed!

**Comment by jliexpensify** [comment]
Deployed to prod as of last week so going to close this out. Rory - feel ree to re-open if you think we need to keep tabs on this GH.
